mitol-django-mail/mitol/mail/api.py

```diff
@@ -0,0 +1,247 @@
+"""
+Email APIs
+
+Example usage:
+
+# get recipients
+recipients = User.objects.all()[:10]
+
+with get_message_sender(TestMessage) as sender:
+    sender.build_and_send_message(user=user)
+"""
+import contextlib
+import logging
+import re
+from copy import deepcopy
+from os import path
+from types import SimpleNamespace
+from typing import Generator, Iterable, Optional, Tuple, Type, Union
+
+import premailer
+from bs4 import BeautifulSoup
+from django.conf import settings
+from django.contrib.auth.base_user import AbstractBaseUser
+from django.core.exceptions import ImproperlyConfigured
+from django.core.mail import get_connection as django_get_connection
+from django.core.mail.backends.base import BaseEmailBackend
+from django.template.loader import render_to_string
+from django.utils.module_loading import import_string
+
+from mitol.mail.messages import TemplatedMessage
+
+
+log = logging.getLogger(__name__)
+
+
+def get_message_classes() -> Iterable[Type[TemplatedMessage]]:
+    classes = []
+    for message_cls in getattr(settings, "MITOL_MAIL_MESSAGE_CLASSES", []):
+        try:
+            classes.append(import_string(message_cls))
+        except ImportError:
+            raise ImproperlyConfigured(
+                f"MITOL_MAIL_MESSAGE_CLASSES is invalid, could not import: {message_cls}"
+            )
+    return classes
+
+
+def safe_format_recipient(
+    recipient_or_user: Union[AbstractBaseUser, str]
+) -> Optional[str]:
+    """
+    Returns a "safe"formatted recipient
+    This means if MAILGUN_RECIPIENT_OVERRIDE is set, we only use that
+
+    Args:
+        recipient_or_user (str or AbstractBaseUser): recipient user or email str
+
+    Returns:
+        str or None: the formatted email str
+    """
+    override_recipient = getattr(settings, "MITOL_MAIL_RECIPIENT_OVERRIDE", None)
+    format_func_name = getattr(
+        settings,
+        "MITOL_MAIL_FORMAT_RECIPIENT_FUNC",
+        "mitol.mail.defaults.format_recipient",
+    )
+    format_func = import_string(format_func_name)
+
+    # we set this for local development so we don't actually email someone
+    if override_recipient is not None:
+        return override_recipient
+
+    if isinstance(recipient_or_user, AbstractBaseUser):
+        return format_func(recipient_or_user)
+
+    return recipient_or_user
```

Comment:
It seems like we should be able to implement this and `can_email_user` without checking these settings and importing functions every time we call it. This might be too hacky, but one option would be to use the app config:

```python
class MailApp(AppConfig):
    # ...
    format_recipient = none
    def ready(self) -> None:
        _format_recipient_fn_name = getattr(
            settings,
            "MITOL_MAIL_FORMAT_RECIPIENT_FUNC",
            "mitol.mail.defaults.format_recipient",
        )
        self.format_recipient = import_string(_format_recipient_fn_name)

# Usage
from django.apps import apps
apps.get_app_config('mail').format_recipient(user)
```

Reply:
So I actually went down a similar route initially, if I recall correctly it made testing this stuff very difficult and introduced a mess of circular dependencies (hence why the import is inlined). Under the hood this calls `importlibs.import_module`, which is cached by the Python interpreter so there's not a performance concern about the inlining either.